### ISSUE TITLE
Is it more reasonable to return 'Service Unavailable' than Internal Server Error when the server is overloaded?

### ISSUE THREAD

**Issue by Felix-zhoux** [scylladb/scylladb#16262]
In the alternator interface, when the current server is under heavy pressure, some requests may cause timeout exception errors. At this time, returning `http code 503 Service Unavailable` will make it easier for the client to perceive what is happening compared to `http code 500 Internal Server Error`. Compared with the service quality after the load is reduced, It can heal itself. In other words, service unavailability is temporary.

ref:https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.MessagesAndCodes:~:text=Service%20Unavailable%20(HTTP%20503)

**Comment by nyh** [comment]
I agree that the "500 Internal Server Error" seems very vague (can mean a lot of different things), and perhaps even scary, but I also want us to consider the following: 

The DynamoDB's documentation https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.MessagesAndCodes
suggests that both for Internal Server Error (500) and Service Unavailable (503), the client's behavior should be the same - to retry. However, the document makes an interesting comment only about Internal Server Error:

> You might encounter internal server errors while working with items. These are expected during the lifetime of a table. Any failed requests can be retried immediately.
> When you receive a status code 500 on a write operation, the operation may have succeeded or failed. If the write operation is a TransactWriteItem request, then it is OK to retry the operation. If the write operation is a single-item write request such as PutItem, UpdateItem, or DeleteItem, then your application should read the state of the item before retrying the operation, and/or use [Condition expressions](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html) to ensure that the item remains in a correct state after retrying regardless of whether the prior operation succeeded or failed. If idempotency is a requirement for the write operation, please use [TransactWriteItem](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transaction-apis.html#transaction-apis-txwriteitems), which supports idempotent requests by automatically specifying a ClientRequestToken to disambiguate multiple attempts to perform the same action.

In other words, the "503 Service Unavailable" error suggests the service was completely unavailable - your request was completely ignored and never done, and you should retry it, whereas "Internal Server Error" says that there was an error **in the middle** of processing your request, and it is not known whether your request succeeded or not.

Our timeout errors are more like Internal Server Error in this sense - it is possible (though not very likely) that the user received a timeout on a write when it has actually already succeeded, or in some way partially succeeded. If this is the case, it's not clear that 503 Service Unavailable is the right message.

In the past there has been some effort in Scylla to do "load shedding" in interactive workloads (see https://github.com/scylladb/scylladb/pull/8680): When the system is overloaded with interactive (unlimited concurrency) requests, don't continue fail requests after working on them for 2 seconds, but rather fail them immediately, with a separate "OVERLOADED" error. Such a request was never started, and can reasonably return 503 Service Unavailable in Alternator.

Note that I'm not completely shooting down your idea - I'm not the foremost expert in load shedding or overload handling, and I'm open to hear your idea and reasoning why you think your idea of converting timeouts to 503 errors is a good one.
